Round of fixes for functional collectives

kumpera opened the pull request:
Move collective registration to torch.__init__ to handle multipy warmup. 
Fix all_reduce with non-contiguous tensors.

wconstab commented on torch/__init__.py:
```diff
@@ -1489,3 +1489,10 @@ def _sparse_coo_tensor_unsafe(*args, **kwargs):
+# Trigger registration of functional distributed collectives
+# This is needed by torch_deploy
+try:
+    import torch.distributed._functional_ops as funops
```
do you want to bind these to a name (funops seems a bit playful) or do you just want to `import torch.distributed._functional_ops` here and not expose them under a short name

kumpera replied:
I'll `del funops` to ensure nothing gets imported into the `torch` module..

wconstab commented on torch/__init__.py:
```diff
@@ -1489,3 +1489,10 @@ def _sparse_coo_tensor_unsafe(*args, **kwargs):
+# This is needed by torch_deploy
+try:
+    import torch.distributed._functional_ops as funops
+except BaseException as _:
```
So this isn't gonna be an ImportError? why BaseException

ezyang replied:
Don't suppress like this. Figure out the condition in which this module can be imported and do it that way

kumpera replied:
👍

kumpera replied:
Removed the import from `torch`

wconstab commented on torch/distributed/_functional_collectives.py:
```diff
@@ -145,36 +147,25 @@ def _all_reduce(self, reduceOp, tag, ranks, group_size):
     assert group is not None
 
-    inplace_tensor = self.clone()
+    inplace_tensor = self.clone(memory_format=torch.contiguous_format)
```
i _think_ the inductor version is already ok,  since it just creates a new 'FlexibleLayout' for AllReduce and doesn't do anything morally equivalent of '.clone'

kumpera replied:
We ran into this issue when running our e2e model.

wconstab reviewed (approved):
seems ok to me, tagging @ezyang to take a look too

ezyang reviewed (dismissed):
try except BaseException suppress is a big no no

Skylion007 commented on torch/__init__.py:
```diff
@@ -1489,3 +1489,11 @@ def _sparse_coo_tensor_unsafe(*args, **kwargs):
+try:
+    import torch.distributed._functional_ops as funops
+    del funops
+except ImportError as _:
```
```suggestion
except ImportError:
```

kumpera replied:
I moved away from changing torch.__init__

kumpera committed:
[PTD] Round of fixes for functional collectives.

Make op registration lazy if running under torch_deploy.
Fix all_reduce with non-contiguous tensors.
Assert that _all_gather_into_tensor usage of new_empty produces a contiguous tensor.

kumpera committed:
Remove lazy registration and emit warning under torch::deploy.:

wconstab reviewed (approved):
ok looks good, we can unblock deploy later

kumpera commented:
@pytorchmergebot merge
